groups/bsl/bslstl/bslstl_variant.h

```diff
@@ -0,0 +1,3338 @@
+}
+#else  //BSL_VARIANT_FULL_IMPLEMENTATION
+
+template <class R, class VISITOR, class VARIANT>
+R visit(VISITOR& Visitor, VARIANT& Variant)
+{
+    return BloombergLP::bslstl::VisitImpl<R>(Visitor, Variant);
+}
+
+template <class R, class VISITOR, class VARIANT>
+R visit(VISITOR& Visitor, BloombergLP::bslmf::MovableRef<VARIANT> Variant)
+{
+    VARIANT& lvalue = Variant;
+
+    return BloombergLP::bslstl::MoveVisitImpl<R>(Visitor, lvalue);
+}
+
+template <class VISITOR, class VARIANT>
+typename bsl::invoke_result<
+    VISITOR,
+    typename bsl::variant_alternative<0, VARIANT>::type&>::type
+visit(VISITOR& Visitor, VARIANT& Variant)
+{
+    typedef typename bsl::invoke_result<
+        VISITOR,
+        typename bsl::variant_alternative<0, VARIANT>::type&>::type R;
+    return BloombergLP::bslstl::VisitImpl<R>(Visitor, Variant);
+}
+
+template <class VISITOR, class VARIANT>
+typename bsl::invoke_result<
+    VISITOR,
+    BloombergLP::bslmf::MovableRef<
+        typename bsl::variant_alternative<0, VARIANT>::type> >::type
+visit(VISITOR& Visitor, BloombergLP::bslmf::MovableRef<VARIANT> Variant)
+{
+    VARIANT& lvalue = Variant;
+
+    typedef typename bsl::invoke_result<
+        VISITOR,
+        BloombergLP::bslmf::MovableRef<
+            typename bsl::variant_alternative<0, VARIANT>::type> >::type R;
+    return BloombergLP::bslstl::MoveVisitImpl<R>(Visitor, lvalue);
+}
+
+#endif  // BSL_VARIANT_FULL_IMPLEMENTATION
+
+}  // close namespace bsl
+
+namespace BloombergLP {
+namespace bslstl {
+
+                          // ========================
+                          // class Variant_NoSuchType
+                          // ========================
+
+struct Variant_NoSuchType {
+    // This component-private trivial tag type is used to distinguish between
+    // arguments passed by a user, and an 'enable_if' argument.  It is not
+    // default constructible so the following construction never invokes a
+    // constrained single parameter constructor:
+    //..
+    //   variant<SomeType> o(int, {});
+    //..
+
+    // CREATORS
+    explicit BSLS_KEYWORD_CONSTEXPR Variant_NoSuchType(
+                                                    int) BSLS_KEYWORD_NOEXCEPT;
+        // Create an 'Variant_NoSuchType' object.  Note that the argument is
+        // not used.
+};
+
+// CREATORS
+inline
+BSLS_KEYWORD_CONSTEXPR Variant_NoSuchType::Variant_NoSuchType(
+                                                     int) BSLS_KEYWORD_NOEXCEPT
+{
+    // This 'BSLS_KEYWORD_INLINE_CONSTEXPR' function has to be defined before
+    // initializing the 'BSLS_KEYWORD_INLINE_CONSTEXPR' value, 'NoSuchType',
+    // below.
+}
+
+#ifdef BSLS_COMPILERFEATURES_SUPPORT_CONSTEXPR
+BSLS_KEYWORD_INLINE_CONSTEXPR Variant_NoSuchType varNoSuchType =
+    Variant_NoSuchType(0);
+#else   //BSLS_COMPILERFEATURES_SUPPORT_CONSTEXPR
+extern const Variant_NoSuchType varNoSuchType;
+#endif  //BSLS_COMPILERFEATURES_SUPPORT_CONSTEXPR
```

Comment:
oh, i forgot to put a definition in bslstl_variant.cpp :/
Darn on the C++11/14
Sure, your suggestion is better and also simpler.
Same pattern exists in bslstl_optional (minus the missing definition). Shall I change it there too ?

Reply:
I looked in `bslstl_optional.h` and it appears that you use it in two ways there: for `optNoSuchType` and `nullopt`.

In the case of `optNoSuchType` I do think it should be fixed although I think you should do that in a separate PR.

In the case of `nullopt` it's a bit trickier because we do need `bsl::nullopt` to exist as part of the interface; it can't simply be eliminated the way `optNoSuchType` can. I don't think there's anything we can do about it at this point; the only way to give it external linkage would be to make it not `constexpr`, and it seems too late to make such a change. We'll just have to be really careful when using it.

Reply:
sure, I'll add this to the list of optional things to fix.

I don't think it's really an issue we need to worry about. These tag types don't really have a value - even if different copies of `nullopt` do get used, I can't think of anything that can go wrong.